scheduler-pool/src/lib.rs

```diff
@@ -0,0 +1,442 @@
+        if result.is_ok() || !fail_fast {
+            *result = execute_batch(
+                &batch_with_indexes,
+                context.bank(),
+                self.pool.transaction_status_sender.as_ref(),
+                self.pool.replay_vote_sender.as_ref(),
+                timings,
+                self.pool.log_messages_bytes_limit,
+                &self.pool.prioritization_fee_cache,
+            );
+        }
+    }
+
+    fn schedule_termination(&mut self) {
+        drop::<Option<SchedulingContext>>(self.context.take());
+    }
+
+    fn wait_for_termination(&mut self, wait_reason: &WaitReason) -> Option<ResultWithTimings> {
+        let keep_result_with_timings = match wait_reason {
+            WaitReason::ReinitializedForRecentBlockhash => {
+                // rustfmt...
+                true
+            }
+            WaitReason::TerminatedToFreeze
+            | WaitReason::TerminatedFromBankDrop
+            | WaitReason::TerminatedInternallyByScheduler => false,
+        };
+
+        self.schedule_termination();
+
+        // current simplest form of this trait impl doesn't block the current thread materially
+        // just with the following single mutex lock. Suppose more elaborated synchronization
+        // across worker threads here in the future...
+
+        if keep_result_with_timings {
+            None
+        } else {
+            self.result_with_timings
+                .lock()
+                .expect("not poisoned")
+                .take()
+        }
+    }
+
+    fn scheduling_context(&self) -> Option<&SchedulingContext> {
+        self.context.as_ref()
+    }
+
+    fn replace_scheduler_context(&mut self, context: SchedulingContext) {
+        self.context = Some(context);
+        *self.result_with_timings.lock().expect("not poisoned") = None;
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use {
+        super::*,
+        crate::SchedulerPool,
+        assert_matches::assert_matches,
+        solana_runtime::{
+            bank::Bank,
+            bank_forks::BankForks,
+            genesis_utils::{create_genesis_config, GenesisConfigInfo},
+            installed_scheduler_pool::SchedulingContext,
+            prioritization_fee_cache::PrioritizationFeeCache,
+        },
+        solana_sdk::{pubkey::Pubkey, signer::keypair::Keypair, system_transaction},
+        std::sync::Arc,
+    };
+
+    #[test]
+    fn test_scheduler_pool_new() {
+        solana_logger::setup();
+
+        let _ignored_prioritization_fee_cache = Arc::new(PrioritizationFeeCache::new(0u64));
+        let pool = SchedulerPool::new_dyn(None, None, None, _ignored_prioritization_fee_cache);
+
+        // this indirectly proves that there should be circular link because there's only one Arc
+        // at this moment now
+        assert_eq!((Arc::strong_count(&pool), Arc::weak_count(&pool)), (1, 1));
+        let debug = format!("{pool:#?}");
+        assert!(!debug.is_empty());
```

Comment:
Wouldn't this always be true?
`pool` is an `Arc<SchedulerPool>`.
`SchedulerPool` is a struct with a default implementation for `Debug`.
And `Arc` [forwards its `Debug` implementation to the contained type](https://doc.rust-lang.org/src/alloc/sync.rs.html#2458).

So `debug` will contain a debug output for a `SchedulerPool` instance.  Which should never be an empty string.

Reply:
yep, this is done for sanity check and coverage of Debug impl. i did similar thing Bank. maybe i need to comment both as such